File: src/main/java/simplifying/conditional/expressions/AuthorizationService.java
package simplifying.conditional.expressions;

public class AuthorizationService {
    private static final int MAJORITY = 18;

    public boolean isAuthorized(User user, String action) {
        return isActionValid(action)
                && isUserAuthorized(user);
    }

    private static boolean isActionValid(String action) {
        return action != null && !action.equals("");
    }

    private boolean isUserAuthorized(User user) {
        return user.getAge() >= MAJORITY
                && !user.isDisabled()
                && user.isLoyal();
    }
}

File: src/scala/src/main/scala/simplifying/conditional/expressions/AuthorizationService.scala
package simplifying.conditional.expressions

class AuthorizationService {
  private val Majority = 18

  def isAuthorized(user: Option[User], action: Option[String]): Boolean = {
    isActionValid(action) && canAuthorizeUser(user)
  }

  private def isActionValid(action: Option[String]) = action.exists(_.nonEmpty)

  private def canAuthorizeUser(user: Option[User]): Boolean = user.exists(u => u.age > Majority && !u.isDisabled && u.isLoyal)
}
